packages/environment-glimmerx/-private/dsl/integration-declarations.d.ts

```diff
@@ -1,9 +1,122 @@
+// with templates to include the information necessary for Glint to typecheck them.
+import { ComponentLike, HelperLike } from '@glint/template';
+import {
+  BoundModifier,
+  Context,
+  DirectInvokable,
+  EmptyObject,
+  FlattenBlockParams,
+  TemplateContext,
+} from '@glint/template/-private/integration';
+
+//////////////////////////////////////////////////////////////////////
+// Components
+
 import '@glimmerx/component';
-import { TemplateComponent } from '../../component';
+
+import { ExpandSignature } from '@glimmer/component/-private/component';
 
 // Declaring that `hbs` returns a `TemplateComponent` prevents vanilla `tsc` from freaking out when
 // it sees code like `const MyThing: TC<Sig> = hbs...`. Glint itself will never see `hbs` get used, as
 // it's transformed to the template DSL before typechecking.
+type ComponentContext<This, S> = TemplateContext<
+  This,
+  ExpandSignature<S>['Args']['Named'],
+  FlattenBlockParams<ExpandSignature<S>['Blocks']>,
+  ExpandSignature<S>['Element']
+>;
+
 declare module '@glimmerx/component' {
-  export function hbs(source: TemplateStringsArray): TemplateComponent<any>;
+  export default interface Component<S> extends InstanceType<ComponentLike<S>> {
+    [Context]: ComponentContext<this, S>;
+  }
+
+  export interface TemplateComponentInstance<S> extends InstanceType<ComponentLike<S>> {
+    [Context]: ComponentContext<null, S>;
+  }
+}
+
+//////////////////////////////////////////////////////////////////////
+// Helpers
+
+import '@glimmerx/helper';
+
+type _FnHelper = DirectInvokable<{
+  <Ret, Args extends unknown[]>(args: EmptyObject, f: (...rest: Args) => Ret): (
+    ...rest: Args
+  ) => Ret;
+  <A, Ret, Args extends unknown[]>(args: EmptyObject, f: (a: A, ...rest: Args) => Ret, a: A): (
+    ...rest: Args
+  ) => Ret;
+  <A, B, Ret, Args extends unknown[]>(
+    args: EmptyObject,
+    f: (a: A, b: B, ...rest: Args) => Ret,
+    a: A,
+    b: B
+  ): (...rest: Args) => Ret;
+  <A, B, C, Ret, Args extends unknown[]>(
+    args: EmptyObject,
+    f: (a: A, b: B, c: C, ...rest: Args) => Ret,
+    a: A,
+    b: B,
+    c: C
+  ): (...rest: Args) => Ret;
+  <A, B, C, D, Ret, Args extends unknown[]>(
+    args: EmptyObject,
+    f: (a: A, b: B, c: C, d: D, ...rest: Args) => Ret,
+    a: A,
+    b: B,
+    c: C,
+    d: D
+  ): (...rest: Args) => Ret;
+  <A, B, C, D, E, Ret, Args extends unknown[]>(
+    args: EmptyObject,
+    f: (a: A, b: B, c: C, d: D, e: E, ...rest: Args) => Ret,
+    a: A,
+    b: B,
+    c: C,
+    d: D,
+    e: E
+  ): (...rest: Args) => Ret;
+  <A, B, C, D, E, G, Ret, Args extends unknown[]>(
+    args: EmptyObject,
+    f: (a: A, b: B, c: C, d: D, e: E, g: G, ...rest: Args) => Ret,
+    a: A,
+    b: B,
+    c: C,
+    d: D,
+    e: E,
+    g: G
+  ): (...rest: Args) => Ret;
+}>;
+
+declare module '@glimmerx/helper' {
+  export interface Helper extends InstanceType<HelperLike> {}
+  export interface FnHelper extends _FnHelper {}
+}
```

Comment:
I think the `Helper<S, S>` issue you're seeing is due to TS losing all sense of reality when you try to do declaration merging on a `type` declaration. Rather than the `type Helper<S>`, I think we need to target the `interface HelperInstance<S>`.

Locally this gets `helper.test.ts` passing for me:
```suggestion
declare module '@glimmerx/helper' {
  export interface FnHelper extends _FnHelper {}
}

declare module '@glimmerx/helper/dist/commonjs/src/helper' {
  export interface HelperInstance<S> extends InstanceType<HelperLike<S>> {}
}
```

Reply:
This definitely did the trick